### PR TITLE
[GL.Rendering3D] Clean up GL components

### PR DESCRIPTION
**Pull request by alxbilger**
- Fix link issues because of missing *_API
- Follow naming guidelines + depreciation of old variables
- OglGrid uses a cached list of vertices
- Use of `saveLastState` and `restoreLastState`

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglSceneFrame.cpp [hunk 1]
```diff
@@ -57,21 +59,15 @@ void OglSceneFrame::reinit()
...
-{
-
-}
-
 void OglSceneFrame::draw(const core::visual::VisualParams* vparams)
 {
-    if (!drawFrame.getValue()) return;
+    if (!d_drawFrame.getValue()) return;
 
     glPushAttrib( GL_ALL_ATTRIB_BITS);
```
Comment (fredroy) [inline SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglSceneFrame.cpp]: You can replace it with the saveLastState()

**Review by fredroy — CHANGES_REQUESTED** [review]

**Comment by fredroy** [comment]
Other than my little contribution, well done 👌

File: SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglSceneFrame.cpp (hunk 1 above)
Comment (alxbilger) [reply SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglSceneFrame.cpp]: Sure. I took the opportunity to use draw tools as much as possible, replacing the corresponding OpenGL calls

**Comment by epernod** [comment]
I see @fredroy you requested changes. I think it is clear now?

File: SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglGrid.cpp [hunk 2]
```diff
@@ -35,17 +35,29 @@ int OglGridClass = core::RegisterObject("Display a simple grid")
...
-    , size(initData(&size, 10.0f,  "size", "Size of the squared grid"))
-    , nbSubdiv(initData(&nbSubdiv, 16,  "nbSubdiv", "Number of subdivisions"))
-    , color(initData(&color, sofa::type::RGBAColor(0.34117647058f,0.34117647058f,0.34117647058f,1.0f),  "color", "Color of the lines in the grid. default=(0.34,0.34,0.34,1.0)"))
-    , thickness(initData(&thickness, 1.0f,  "thickness", "Thickness of the lines in the grid"))
-    , draw(initData(&draw, true,  "draw", "Display the grid or not"))
-{}
+    : d_plane(initData(&d_plane, std::string("z"),  "plane", "Plane of the grid"))
+    , d_size(initData(&d_size, 10.0f,  "size", "Size of the squared grid"))
+    , d_nbSubdiv(initData(&d_nbSubdiv, 16,  "nbSubdiv", "Number of subdivisions"))
+    , d_color(initData(&d_color, sofa::type::RGBAColor(0.34117647058f,0.34117647058f,0.34117647058f,1.0f),  "color", "Color of the lines in the grid. default=(0.34,0.34,0.34,1.0)"))
```
Comment (epernod) [inline SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglGrid.cpp]: 0.34117647058f I think we are more sharp on line color than on CG convergence!

File: SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglGrid.cpp [hunk 3]
```diff
@@ -55,120 +67,133 @@ void OglGrid::reinit()
...
-//    f_bbox.setValue(sofa::type::BoundingBox(min,max));
+    buildGrid();
 
+    //bounding box for the camera
+    auto s = d_size.getValue();
+    sofa::type::Vec3f min,max;
+    switch(internalPlane)
+    {
+        case PLANE_X:
+            min = sofa::type::Vec3f(-s*0.1f, -s*0.5f, -s*0.5f);
```
Comment (epernod) [inline SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglGrid.cpp]: maybe create 2 variables for s*0.1f and s*0.5f , if you think we will change that values some days...

File: SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglGrid.h [hunk 4]
```diff
@@ -37,25 +37,33 @@ class OglGrid : public core::visual::VisualModel
...
     void updateVisual() override;
+    void buildGrid();
+
+    SOFA_ATTRIBUTE_DISABLED__RENDERING3D_DATA_WITH_PREFIX
+    DeprecatedAndRemoved size, nbSubdiv, color, thickness, draw;
 
 protected:
 
+    PLANE internalPlane;
+    sofa::type::vector<Vector3> m_drawnPoints;
```
Comment (epernod) [inline SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglGrid.h]: small doxygen doc, even if the name is crystal clear

File: SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglSceneFrame.cpp [hunk 5]
```diff
@@ -57,229 +58,188 @@ void OglSceneFrame::reinit()
...
+    default:
+        drawArrows(vparams);
+        break;
+
+    case 1:
+        drawCylinders(vparams);
+        break;
+
+    case 2:
+        drawCubeCones(vparams);
```
Comment (epernod) [inline SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglSceneFrame.cpp:184]: what is a cube cone? pyramid?

Comment (epernod) [reply SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglSceneFrame.cpp:184]: If I understood well, you are updating/changing some rendering, maybe a small screenshot just for the record.

**Comment by epernod** [comment]
[ci-build][with-all-tests]

File: SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglSceneFrame.cpp (hunk 5 above)
Comment (alxbilger) [reply SofaKernel/modules/Sofa.GL/Component/Rendering3D/src/sofa/gl/component/rendering3d/OglSceneFrame.cpp:184]: Note that CubeCones already existed before (but I was the author).
I attached a video capture to visualize what CubeCones is.

https://user-images.githubusercontent.com/10572752/178247590-c43c3c32-b626-46a2-bdcf-f8aa817090c3.mp4

The rendering should be the same. Only the way to call it changes.

**Review by fredroy — APPROVED** [review]

**Comment by damienmarchal** [comment]
Hi all, 

Thanks for the PR. 

In the past it was consider that draw tools was a rendering abstraction layer to be used for debugging/inspection or GUI helpers. The recommendation was to use drawTools in the Base::visual() overrides. The main ideas was to allow component developpers to have rendering tools without having a dependency to opengl. On its side, OglXXX component in Sofa.GL exists to hold compoennts that, for whatever reason (performance, lack of time), are still implemented using the opengl api.

Up to now, the evolution guidelines was something like:
- replace in the complete code base ogl call by drawTools calls into xxx::draw() 
- move every component for which ogl call remains (mostly in xxxx:drawVisual) into the Sofa.GL.XXX

This PR does something else, as it replace opengl calls from Sofa.GL.OglXXX by drawTools's ones. Implementing component rendering using drawTools instead of ogl calls is probably desirable (if there is no performance cost)... but it cannot be in Sofa.GL. 

I recommand to create a new component in Sofa.Component.Visual, for the component using drawTools and if needed in Sofa.GL.Component have the ogl version subclassing it (if needed).

**Review by damienmarchal — CHANGES_REQUESTED** [review]
Is there performance cost/benefit for this PR ? 
Please consider moving component not using opengl calls into Sofa.Component.Visual.

**Comment by alxbilger** [comment]
@damienmarchal Thanks for your suggestion. Indeed, some components can now be moved to a generic module not depending on OpenGL 👍 🎊 . But I am confused by your comment. Did you want to suggest other things than just moving the components?

**Comment by damienmarchal** [comment]
@alxbilger 

In addition to have very component not relying anymore on a dependency to opengl should be moved to a generic module. 
I would suggest having a layer of compatibility/deprecation (possibly build on top of the generic module) to to tell users that scenes like:
```console
   RequiredPlugin("Sofa.GL.Component") 
   OglModel()
   OglGrid() 
```
Are still working but displays nice guidance on how to update to more or less something of that kind:
```console
   RequiredPlugin("Sofa.Components.Visual")                 
   VisualModel()                                                          
   VisualGrid() 
```

**Comment by alxbilger** [comment]
@damienmarchal Does https://github.com/sofa-framework/sofa/pull/3115/commits/e8bb1deb711c2f512739a1afa373cbaef96bcb59 match your expectations?

**Pull request merged**